myia/pipeline/steps.py

```diff
@@ -549,6 +549,15 @@ def convert_arg(self, arg, orig_t: AbstractArray, backend):
     return [arg]
 
 
+@overload  # noqa: F811
+def convert_arg(self, arg, orig_t: AbstractUnion, backend):
+    for opt in orig_t.options:
+        try:
+            return self(arg, opt, backend)
+        except TypeError:
+            continue
+
```

Comment:
Two things:
- I don't think that having Union types for top-level arguments is something we want to support.
- If we do keep it, there should be an error when none of the options match.

Reply:
We do want to support this: if the function operates on trees, one of the arguments will be a tree, and a tree (let's say a binary tree) would have the type `Tree :: Union[Nil, Cons[left :: Tree, right :: Tree]]`.

Good catch about the error, though, that's definitely something I forgot.

Reply:
Ok about support.